nucliadb/src/nucliadb/search/search/query.py

```diff
@@ -191,18 +191,45 @@ def _get_query_information(self) -> Awaitable[QueryInfo]:
         return self._query_information_task
 
     async def _query_information(self) -> QueryInfo:
-        vectorset = await self.select_vectorset()
+        vectorset = await self.select_query_vectorset()
         return await query_information(
             self.kbid, self.query, vectorset, self.generative_model, self.rephrase, self.rephrase_prompt
         )
 
+    def _get_vectorset(self) -> Awaitable[Optional[str]]:
+        if self._get_vectorset_task is None:
+            self._get_vectorset_task = asyncio.create_task(self._select_vectorset())
+        return self._get_vectorset_task
+
+    async def _select_vectorset(self) -> Optional[str]:
+        if self.vectorset is not None:
+            return self.vectorset
+
+        # When vectorset is not provided we get the default from Predict API
+
+        query_information = await self._get_query_information()
+
+        if query_information.sentence is None:
+            logger.error(
+                "Asking for a vectorset but /query didn't return one", extra={"kbid": self.kbid}
+            )
+            return self.vectorset
+
+        for vectorset in query_information.sentence.vectors.keys():
```

Comment:
nit: is popitem better/safer?

Reply:
I prefer not to modify the `query_information` object